autodp/mechanism_zoo.py

```diff
@@ -79,40 +117,104 @@ def __init__(self, sigma=None, name='Gaussian'):
+    def __init__(self, b=None, name='Laplace', RDP_off=False, phi_off=True):
+        """
+        b: the ratio of the scale parameter and L1 sensitivity.
+        RDP_off: if False, then we characterize the mechanism using RDP.
+        fdp_off: if False, then we characterize the mechanism using fdp.
+        phi_off: if False, then we characterize the mechanism using phi-function.
+        """
         Mechanism.__init__(self)
 
         self.name = name
         self.params = {'b': b} # This will be useful for the Calibrator
+
         self.delta0 = 0
-        if b is not None:
+        if not phi_off:
+
+            log_phi_p = lambda x: phi_bank.phi_laplace(self.params, x)
+            log_phi_q = lambda x: phi_bank.phi_laplace(self.params, x)
+
+            self.log_phi_p = log_phi_p
+            self.log_phi_q = log_phi_q
+            self.propagate_updates((log_phi_p, log_phi_q), 'log_phi')
+
+
+        if not RDP_off:
             new_rdp = lambda x: rdp_bank.RDP_laplace({'b': b}, x)
             self.propagate_updates(new_rdp, 'RDP')
 
 
+
 class RandresponseMechanism(Mechanism):
 
     """
-        param params:
-        'p' --- is the Bernoulli probability p of outputting the truth.
+    The randomized response mechanism that supports RDP and phi-function based characterization.
+    TODO: assert when p is None.
+    """
+    def __init__(self, p=None, RDP_off=False, phi_off=True, name='Randresponse'):
+        """
+        p: the Bernoulli probability p of outputting the truth.
         """
-
-    def __init__(self, p=None, name='Randresponse'):
         Mechanism.__init__(self)
 
         self.name = name
-        self.params = {'p': p}  # This will be useful for the Calibrator
+        self.params = {'p': p}
         self.delta0 = 0
-        if p is not None:
+
+        if not RDP_off:
             new_rdp = lambda x: rdp_bank.RDP_randresponse({'p': p}, x)
             self.propagate_updates(new_rdp, 'RDP')
             approxDP = lambda x: dp_bank.get_eps_randresp_optimal(p, x)
             self.propagate_updates(approxDP, 'approxDP_func')
 
+        if not phi_off:
+            log_phi = lambda x: phi_bank.phi_rr_p({'p': p, 'q':1-p}, x)
+            self.log_phi_p = self.log_phi_q = log_phi
+            self.propagate_updates((self.log_phi_p, self.log_phi_q), 'log_phi')
+
+class zCDP_Mechanism(Mechanism):
+    def __init__(self,rho,xi=0,name='zCDP_mech'):
+        Mechanism.__init__(self)
+
+        self.name = name
+        self.params = {'rho':rho,'xi':xi}
+        new_rdp = lambda x: rdp_bank.RDP_zCDP(self.params, x)
+
+        self.propagate_updates(new_rdp,'RDP')
+
+
+class ExponentialMechanism(zCDP_Mechanism):
+    def __init__(self, eps,RDP_off=False, phi_off=True,  name='ExpMech'):
+        zCDP_Mechanism.__init__(self, eps**2/8, name=name)
+        # the zCDP bound is from here: https://arxiv.org/pdf/2004.07223.pdf
+
+        self.eps_pureDP = eps
+        self.propagate_updates(eps, 'pureDP')
+
+        if not RDP_off:
```

Comment:
This is already done via "propagate_udpate" in pureDP above?

Reply:
Also the choice of eps/2 seems incorrect. While EM satisfies eps^2/8 zCDP, it only satisfies eps-DP rather than eps/2-DP.

Reply:
'''new_rdp = lambda x: rdp_bank.RDP_pureDP({'eps': eps/2.}, x)'''
should be 
```
new_rdp = lambda x: rdp_bank.RDP_pureDP({'eps': eps}, x)
```